kernel/mellanox_ethernet/src/command_queue.rs

```diff
@@ -0,0 +1,740 @@
+    pub fn create(entries: BoxRefMut<MappedPages, [CommandQueueEntry]>, num_cmdq_entries: usize) -> Result<CommandQueue, &'static str> {
+        
+        // initially, all command entries are available
+        let mut available_entries = Vec::with_capacity(num_cmdq_entries);
+        for _ in 0..num_cmdq_entries { available_entries.push(true); }
+
+        // start off by pre-allocating one page for input and output mailboxes per entry
+        let mut mailbox_buffers_input = Vec::with_capacity(num_cmdq_entries);
+        let mut mailbox_buffers_output = Vec::with_capacity(num_cmdq_entries);
+        for _ in 0..num_cmdq_entries {
+            let (mailbox_mp, mailbox_pa) = create_contiguous_mapping(PAGE_SIZE, NIC_MAPPING_FLAGS)?;
+            mailbox_buffers_input.push(vec!((mailbox_mp, mailbox_pa)));
+
+            let (mailbox_mp, mailbox_pa) = create_contiguous_mapping(PAGE_SIZE, NIC_MAPPING_FLAGS)?;
+            mailbox_buffers_output.push(vec!((mailbox_mp, mailbox_pa)));
+        }
+
+        // Assign a random number to token, another OS (Snabb) starts off with 0xAA
+        let token = 0xAA;
+
+        Ok(CommandQueue{ entries, available_entries, token, mailbox_buffers_input, mailbox_buffers_output })
+    }
+
+    /// Find an command queue entry that is not in use
+    fn find_free_command_entry(&self) -> Option<usize> {
+        self.available_entries.iter().position(|&x| x == true)
+    }
+
+    /// Fill in the fields of a command queue entry.
+    /// At the end of the function, the command is ready to be posted using the doorbell in the initialization segment. 
+    /// Returns an error if no entry is available to use.
+    ///
+    /// # Arguments
+    /// * `opcode`: opcode for command that the driver wants to execute
+    /// * `opmod`: opcode modifer, only applicable for certain commands
+    /// * `allocated_pages`: physical address of pages that need to be passed to the NIC. Only used in the MANAGE_PAGES command.
+    pub fn create_command(
+        &mut self, 
+        opcode: CommandOpcode, 
+        opmod: Option<u16>, 
+        allocated_pages: Option<Vec<PhysicalAddress>>,
+    ) -> Result<usize, &'static str> 
+    {
+        let entry_num = self.find_free_command_entry().ok_or("No command entry available")?; 
+        
+        // clear the fields of the command
+        let mut cmdq_entry = CommandQueueEntry::default();
+
+        // set the type of transport and token
+        cmdq_entry.set_type_of_transport();
+        cmdq_entry.set_token(self.token);
+
+        // set the ownership bit to HW-owned.
+        // This bit will be cleared when the command is complete.
+        cmdq_entry.change_ownership_to_hw();
+
+        match opcode {
+            CommandOpcode::EnableHca => {
+                cmdq_entry.set_input_length_in_bytes(12);
+                cmdq_entry.set_output_length_in_bytes(8);
+                cmdq_entry.set_input_inline_data(opcode, opmod, None, None);
```

Comment:
For each one of these `cmdq_entry.set...` functions, there's probably a better way to represent it than just a bunch of constants, it's a bit hard to understand the logic. For example, a table or enum of length values per command opcode, perhaps a function implemented on the `CommandOpcode` enum.

Reply:
something something "rule of representation", as Lin would say :)